### DISCUSSION TITLE
ZKSync Prover Reference Material | Understanding Crate Used in ZKSync Prover

### DISCUSSION THREAD

**Discussion by sarthak608** [zkSync-Community-Hub/zksync-developers#830] [Dev Tooling] [answered]
### Team or Project

Tevaera

### Environment

Testnet

### Select the Dev Tool you are using

Local Node

### Provide a brief description of the functionality you're trying to implement and the issue you are running into.

<img width="539" alt="Screenshot 2024-12-13 at 8 06 11 PM" src="https://github.com/user-attachments/assets/a43b85f1-c62c-4fdd-a080-e1995b329353" />

I am currently working on understanding ZKSync Prover. Could you provide links and other relevant details that could help me gain a deeper understanding of these crates?

**Comment by alhenaai** [comment]
To gain a deeper understanding of the ZKsync Prover and its components, you can refer to the following resources:

1. **ZKsync Prover Overview**: This document provides a comprehensive overview of the ZKsync Prover subsystem, detailing its components and functionality. It's a great starting point to understand the proof generation process in ZKsync.
   - [ZKsync Prover Overview](https://docs.zksync.io/zk-stack/components/prover)

2. **ZKsync Protocol Circuits**: This section explains the role of circuits in the ZKsync proving process, which is crucial for understanding how computations are verified in ZKsync.
   - [How ZKsync Utilizes ZK Proofs](https://docs.zksync.io/zksync-protocol/circuits#how-zksync-utilizes-zk-proofs)
   - [The Proving Process](https://docs.zksync.io/zksync-protocol/circuits#the-proving-process)
   - [Understanding Circuits in ZKsync](https://docs.zksync.io/zksync-protocol/circuits#understanding-circuits-in-zksync)

These resources should provide you with a solid foundation on the ZKsync Prover and its associated technologies. For more in-depth discussions and support, you can also visit the ZKsync developers discussions on Github: [ZKsync Developers Discussions](https://github.com/zkSync-Community-Hub/zksync-developers/discussions).

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by dutterbutter** [comment]
Hey @sarthak608 

The resources mentioned above are the best places to start:

- [ZKsync Prover Overview](https://docs.zksync.io/zk-stack/components/prover)
- [How ZKsync Utilizes ZK Proofs](https://docs.zksync.io/zksync-protocol/circuits#how-zksync-utilizes-zk-proofs)
- [The Proving Process](https://docs.zksync.io/zksync-protocol/circuits#the-proving-process)
- [Understanding Circuits in ZKsync](https://docs.zksync.io/zksync-protocol/circuits#understanding-circuits-in-zksync)
- [Prover Book](https://matter-labs.github.io/zksync-era/prover/latest/)

If you are trying to understand the crates, I definitely suggest reading the implementation. If you have a specific question related to prover please feel free to open a separate discussion post.

**Reply by sarthak608** [reply]
I've read through the available resources, but I haven't found much detailed information regarding the crate itself. I would appreciate your help in guiding me on how to understand the crate used by ZKSync Prover, as some crate lack in-depth explanations on readMe such as bellman , franklin-crypto and others

**Reply by dutterbutter** [reply]
I will reach out to our prover team to see if there is any other material outside of what I have already linked but I am not really sure what kind of resource you are looking for? Can you provide more clarity here?

The best way to understand the crates is to read the code implementation and comments. If you have a specific question that would make it much easier to help here.

**Reply by popzxc** [reply, marked as answer]
If you want to get familiar with lower-level building primitives (boojum/franklin-crypto/etc), then I guess the best way would be to read the following:
- [Read parts 2.13 - 2.15 in the core book](https://matter-labs.github.io/zksync-era/core/v25.2.0/guides/advanced/13_zk_intuition.html)
- [Read the circuits section in the docs](https://docs.zksync.io/zksync-protocol/circuits)
- And from there, dive into the code of the crypto logic. We have three main monorepos:
  - [zksync-crypto](https://github.com/matter-labs/zksync-crypto): primitive libraries for the protocol logic (including pairing/bellman/boojum)
  - [zksync-protocol](https://github.com/matter-labs/zksync-protocol): protocol logic, including the implementation of VM and circuits for it.
  - [zksync-crypto-gpu](https://github.com/matter-labs/zksync-crypto-gpu): Logic for implementing proving on GPU

Unfortunately, right now we don't have detailed information on each particular crate, so reading source code would be the best approach.
